### PR TITLE
feat: add `filter` option to sources

### PR DESCRIPTION
**Pull request by jonatan-branting**
This allows the user to specify a `filter` function for each source.

Solves:
https://github.com/hrsh7th/nvim-cmp/issues/1036
https://github.com/hrsh7th/nvim-cmp/issues/1041
https://github.com/hrsh7th/nvim-cmp/issues/806

@hrsh7th do you see any problems with this implementation or approach in general?

An example (Updated):
```lua
-- don't show entries with kind "Text" from the "nvim_lsp" source
sources = {
{
  name = "nvim_lsp",
  entry_filter = function(entry, ctx)
    local kind = types.lsp.CompletionItemKind[entry:get_kind()]

    if kind == "Text" then return false end
    return true
  },
}
```

By utilizing the `ctx` parameter, the user can also ignore certain
entries in certain contexts.

Notes:
- <s>The name `filter` might not be entirely clear; what does returning `true` or `false` actually result in? Ideas?</s> Update: Changed name to `entry_filter`, and swapped true/false, making it function more like traditional filter functions.
- Documentation has to be added.
- <s>Specs have to be added.</s> There doesn't seem to be a precedence for testing other configuration options, not sure it's entirely necessary either.

### CODE REVIEW

**Comment by Weissle** [comment]
As a user, I think this feature is useful. I have luasnip and some LSP servers also provide snippets. It is annoying that LSP's snippets are always standing before the luasnip's but the quality of them is not so good. 
I know how to adjust their priority, but a filter function provides great scalability.

**Comment by hrsh7th** [comment]
Yes. This feature is useful. I'm thinking about the API design...

**Comment by Massolari** [comment]
Usually, in `filter` functions, when you return `true` that means you want the item, if you don't want the item, you return `false` .

File: lua/cmp/source.lua
Comment (bennypowers) [inline lua/cmp/source.lua:120]:
```suggestion
      if type(filter) ~= 'function' or filter(e, ctx) then
```

if no filter was defined, insert
otherwise, if the filter was defined and it passes for the entry/context pair, insert

This could also be done by defaulting the filter to `() -> true`

Comment (gotgenes) [reply lua/cmp/source.lua:120]: >This could also be done by defaulting the filter to `() -> true`

+1 to this approach! I find setting a default filter of `() -> true` and then always applying the filter easier to understand. [Here's a talk on this approach](https://www.youtube.com/watch?v=AnZ0uTOerUI) by Michael Feathers.

Comment (jonatan-branting) [reply lua/cmp/source.lua:120]: I do like that approach better. I'd implement it, but I got the feeling that @hrsh7th cooking up something on his own.

Comment (Weissle) [reply lua/cmp/source.lua:120]: @hrsh7th seems not active recently.
But it would be better to have a discuss with he to avoid you are working on the same feature.

Comment (jonatan-branting) [reply lua/cmp/source.lua]: See: [03a7a91](https://github.com/hrsh7th/nvim-cmp/pull/1067/commits/03a7a91c7e1da6ce0147942a91b08fe3c52884d9)

**Comment by jonatan-branting** [comment]
I pushed some updates according to feedback ([03a7a91](https://github.com/hrsh7th/nvim-cmp/pull/1067/commits/03a7a91c7e1da6ce0147942a91b08fe3c52884d9)). If anyone used my fork then this will have broken your configuration, sorry about that! @kkharji and @bennypowers

**Comment by jonatan-branting** [comment]
@hrsh7th if this is good enough for you I'll add documentation and we can maybe get this merged? :)

File: lua/cmp/source.lua
Comment (jonatan-branting) [inline lua/cmp/source.lua:119]: Move this out of the loop, fetching the filter for each iteration seems a tad unnecessary.

File: lua/cmp/source.lua
Comment (bennypowers) [inline lua/cmp/source.lua:245]: nice

**Comment by Weissle** [comment]
What about the document? @hrsh7th may not merge this PR without the document updated.

**Comment by jonatan-branting** [comment]
> What about the document? @hrsh7th may not merge this PR without the document updated.

I'm holding off on that until I get the OK from @hrsh7th whether or not this is the solution he wants to go for or not.

**Comment by hrsh7th** [comment]
I thought it would be more versatile to add an `override` option to the source and add the ability to override all source functions. As shown below.

```lua
{
  name = 'buffer',
  override = {
    complete = function(source_complete, params, callback)
      source_complete(params, function(response))
        -- modify complete response from source.
        callback(modify(response))
      end)
    end
  }
}
```

However, this API is too advanced.
Therefore, I can accept this feature.

Translated with www.DeepL.com/Translator (free version)

**Comment by hrsh7th** [comment]
Please update the docs.

**Comment by jonatan-branting** [comment]
> Please update the docs.

I'll most likely be able to get this done tomorrow :)

File: doc/cmp.txt
Comment (bennypowers) [inline doc/cmp.txt:572]:
```suggestion
      return kind ~= 'Text'
```

File: doc/cmp.txt
Comment (bennypowers) [inline doc/cmp.txt:573]:
```suggestion
```

File: doc/cmp.txt
Comment (bennypowers) [inline doc/cmp.txt:574]:
```suggestion
```

**Comment by jonatan-branting** [comment]
@hrsh7th I've added documentation, let me know if there is anything you want me to change!

File: doc/cmp.txt
Comment (jonatan-branting) [reply doc/cmp.txt]: That's some lightning-quick reviewing, thanks!

**Review by bennypowers — APPROVED** [review]
:heart:

**Comment by bennypowers** [comment]
add this to line 155 in lua/types/cmp.lua

```lua
---@field public entry_filter function(entry: cmp.Entry, ctx: cmp.Context): boolean|nil
```

full block:
```lua
---@class cmp.SourceConfig
---@field public name string
---@field public option table|nil
---@field public priority integer|nil
---@field public trigger_characters string[]|nil
---@field public keyword_pattern string|nil
---@field public keyword_length integer|nil
---@field public max_item_count integer|nil
---@field public group_index integer|nil
---@field public entry_filter function(entry: cmp.Entry, ctx: cmp.Context): boolean|nil
```

File: doc/cmp.txt
Comment (hrsh7th) [inline doc/cmp.txt:572]: It's bit redundunt. Could you change as this?
```
return require('cmp').lsp.CompletionItemKind.Text ~= entry:get_kind()
```

File: lua/cmp/types/cmp.lua
Comment (hrsh7th) [inline lua/cmp/types/cmp.lua:155]: You should add type annotation here.

Comment (jonatan-branting) [reply lua/cmp/types/cmp.lua]: See: [a7d6dd2](https://github.com/hrsh7th/nvim-cmp/pull/1067/commits/a7d6dd29c6ec3acd7bee514c4291aa5a062648f8)

File: doc/cmp.txt
Comment (jonatan-branting) [reply doc/cmp.txt]: See: [a7d6dd2](https://github.com/hrsh7th/nvim-cmp/pull/1067/commits/a7d6dd29c6ec3acd7bee514c4291aa5a062648f8)

**Comment by Weissle** [comment]
Could you take a look? This featue have been backlogged for a month. @hrsh7th

**Review by Weissle — APPROVED** [review]

**Comment by hrsh7th** [comment]
LGTM. Thank you.

**Pull request merged**
